Use the result type for repo methods that throw specific exceptions

Slartibartfass2 opened the pull request:
Closes #291

## What I have made

I recommend reviewing this commit-by-commit.
The refactoring was done in the following steps:
1. Search for repo methods that throw an exception and return a result instead (+ adjust tests) -> first 14 commits
2. Search for service tests that mock exception throwing for methods that do not return a result type, remove them, and return a `Result.Failure` for the other methods. -> following 12 commits
3. In between, use `withUser` instead of `userRepo.getUserById(GrpcContext.getUserIdFromContext()).getOrThrow()` (1fce7514a53e179942509861678ec4b262154f50 and 4a1e7a4bd5b774557b040d28b7e9a901c9cea396)
4. Add some repo helpers (following 3 commits)
5. Add some missing tests (following 2 commits)
6. Adjust imports (last commit)

With this, we could delete ~112 unnecessary tests and a lot of boilerplate code.

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 committed:
refactor: use Result type in ProjectMemberTableRepo

Slartibartfass2 committed:
refactor: use Result type in ProjectPaperTableRepo

Slartibartfass2 committed:
refactor: use common project member check in ReviewService

Slartibartfass2 committed:
refactor: use Result type in ReviewTableRepo

Slartibartfass2 committed:
test: add helper for result type testing

Slartibartfass2 committed:
refactor: add helper for the get[Entity]ById repo methods

Slartibartfass2 committed:
refactor: use Result type in AuthorTableRepo

Slartibartfass2 committed:
refactor: use Result type in CriterionTableRepo

Slartibartfass2 committed:
refactor: use Result type in PaperTableRepo

Slartibartfass2 committed:
refactor: use Result type in ProjectTableRepo

Slartibartfass2 committed:
refactor: use Result type in UserTableRepo

Slartibartfass2 committed:
docs: update comments for refactored methods

Slartibartfass2 committed:
chore: adjust code after rebase

Slartibartfass2 committed:
refactor: don't use getEntityId getters anymore

Previously, we  thought that we require the EntityID object to create a new entity with a reference.
But this isn't the case. The raw value such as a UUID is sufficient.

Slartibartfass2 committed:
refactor: remove obsolete mocking in CriterionService tests

Slartibartfass2 committed:
chore: adjust code after rebase

Slartibartfass2 committed:
refactor: remove obsolete mocking in PaperService

Slartibartfass2 committed:
refactor: use withUser to inject current user

Slartibartfass2 committed:
refactor: remove mocking of retrieving the current user

Slartibartfass2 committed:
test: use specific exceptions in test names

Slartibartfass2 committed:
refactor: use Result type for getProjectPaperByRelativeId

Slartibartfass2 committed:
refactor: remove obsolete mocking in ProjectService

Slartibartfass2 committed:
refactor: remove obsolete mocking in ReadingListService

Slartibartfass2 committed:
refactor: remove obsolete mocking in ReviewService

Slartibartfass2 committed:
refactor: remove obsolete mocking in UserService

Slartibartfass2 committed:
test: use specific exceptions in test names

Slartibartfass2 committed:
refactor: add doesEntityExist repo helper method

Slartibartfass2 committed:
refactor: add getEntities repo helper method

Slartibartfass2 committed:
refactor: use repo helpers

Slartibartfass2 committed:
test: add missing tests for requesting a non-existing project

Slartibartfass2 committed:
test: add tests for withUser helper

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/UpdateUserTest.kt:
```diff
@@ -39,87 +37,54 @@ class UpdateUserTest : MainServiceTest() {
+            coEvery { userRepoMock.getUserById(requestedUserId) } returns Result.success(requestedUser)
 
-            assertThrows<UnauthorizedException.Single> { mainService.updateUser(getExampleRequest()) }
+            assertThrows<UnauthorizedException> { mainService.updateUser(getExampleRequest()) }
```
Since the subtypes only differ in the error message, I think it's fine to assert the parent type.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/UpdateUserTest.kt:
```diff
@@ -39,87 +37,54 @@ class UpdateUserTest : MainServiceTest() {
+        coEvery { userRepoMock.getUserById(otherUser.id) } returns Result.success(otherUser)
 
-        assertThrows<UnauthorizedException.Single> { mainService.updateUser(request) }
+        assertThrows<UnauthorizedException> { mainService.updateUser(request) }
```
Since the subtypes only differ in the error message, I think it's fine to assert the parent type.

Slartibartfass2 committed:
style: use grpc model import aliases consistently

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ProjectMemberTableRepo.kt:
```diff
@@ -31,9 +31,10 @@ import java.util.UUID
     /**
-     * Returns a project member by its composed ID or throws a [NotFoundException] if the project with the passed [projectId] and [userId] doesn't exist.
+     * Returns a [Result] containing the project member by its composed ID or a [NotFoundException] if the project with
+     * the passed [projectId] and [userId] doesn't exist.
```
```suggestion
     * the passed [projectId] or the user with the passed [userId] doesn't exist.
```

Slartibartfass2 replied:
Great suggestion!

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ReviewService.kt:
```diff
@@ -55,56 +56,46 @@ class ReviewService(
+
+            reviews.toGrpcReviews(reviewSelectedCriteriaMap)
         }
 
-        val reviews = repo.getAllReviewsForProjectPaper(projectPaperId)
-        val reviewSelectedCriteriaMap = mutableMapOf<Review, List<String>>()
-        for (review in reviews) {
-            reviewSelectedCriteriaMap[review] = reviewHasCriterionRepo
-                .getSelectedCriteriaIdsForReviewById(review.id).map {
-                    it.toString()
-                }
+    private suspend fun ensureCurrentUserIsProjectMember(projectId: UUID, currentUser: User) {
+        val project = projectRepo.getProjectById(projectId).getOrThrow()
+        val projectMembers = projectMemberRepo.getProjectMembers(project.id)
+        val isProjectMember = projectMembers.any { it.userId == currentUser.id }
+
+        if (!isProjectMember) {
+            verifyServerAdminRole(currentUser) {
+                throw UnauthorizedException.Single(EntityType.PROJECT, project.id.toString(), AccessType.READ, it)
+            }
         }
-        return reviews.toGrpcReviews(reviewSelectedCriteriaMap)
```
Nice refactoring and thank you for your work, otherwise I would have needed to detect this code duplicate in #144.

Slartibartfass2 replied:
I'm happy to help.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/utils/ResultTestHelper.kt:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Assert that the passed [Result] is not a [Result.Failure] and contains the data of type [T].
+ */
+inline fun <reified T> assertResultSuccess(result: Result<T>): T {
```
And maybe also mention, that the data encapsulated in the Result.Sucess is returned, right? Because actually I would expect that this function only asserts something and doesnt return anything.

Slartibartfass2 replied:
Got it. Changed it.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/RepoResultHelper.kt:
```diff
@@ -0,0 +1,44 @@
+import java.util.UUID
+
+/**
+ * Returns an entity by its ID encapsulated by a [Result]. If the entity couldn't be found a [Result.Failure] is
```
```suggestion
 * Returns an entity by its ID, encapsulated by a [Result]. If the entity couldn't be found, a [Result.Failure] is
```
Or am I mistaken? (same would apply for second comment)

Slartibartfass2 replied:
The first comma is not correct, but the second one is.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/RepoResultHelper.kt:
```diff
@@ -0,0 +1,44 @@
+ * @param T The type of the entity.
+ * @param getter The actual getter method that returns a nullable entity.
+ * @param entityType The [EntityType] of the entity.
+ * @param id The ID of the entity.
```
A general question (that must not be resolved in this PR): is it possible to infer the EntityType from the type of the entity T automatically? Because if so, we could simplify some code.

Slartibartfass2 replied:
For this we would have two options:
1. Make a superclass for all entity DTOs and then have a common getter or property for `EntityType`.
2. Enums cannot have a generic type, afaik, and this way we cannot infer a type -> seems impossible

The first option seems more plausible, but in this example it wouldn't help, since the entity is null in the else-block and we couldn't access the EntityType getter or property.

Maybe I'm missing something.

Merseleo replied:
Hm ok. I understand the problem, then lets keep it as it is.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -156,32 +156,27 @@ class UserTableRepo(
+        if (user != null) {
+            Result.success(user)
+        } else {
+            Result.failure(NotFoundException(EntityType.USER, email, identifierType = IdentifierType.EMAIL))
```
I can image that we will face this problem again, that we do not return a basic NotFoundException. What do you think of adding another helper to the RepoResultHelper file: `getEntetityByIdAsResult(getterEntity, getterException)` or something similar?

Slartibartfass2 replied:
I could find a way so that these helpers can be used with any form of key, instead of just IDs.

Merseleo replied:
I think this solution could work well for future calls 👍

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -107,15 +108,15 @@ interface IUserTableRepo {
      * @return The password hash as a [String] for the user with the specified email.
      */
-    suspend fun getPasswordHashByEmail(email: String): String
+    suspend fun getPasswordHashByEmail(email: String): Result<String>
```
You adapted the result type but not the comment as for `getUserById`, is that on purpose? (same for `getUserSettings`)

Slartibartfass2 replied:
No, I just forgot these two methods.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -267,20 +273,30 @@ class UserTableRepo(
+        if (settings != null) {
+            Result.success(settings)
+        } else {
+            Result.failure(NotFoundException(EntityType.USER, id.toString(), identifierType = IdentifierType.ID))
```
I think `IdentifierType.ID` is the default, i.e. it can be omitted and then you can use your helper directly as in `getUserById`.

Slartibartfass2 replied:
Good catch!

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/repository/UserTableRepoTest.kt:
```diff
@@ -74,18 +78,21 @@ class UserTableRepoTest : RepositoryTest(arrayOf(UserTable)) {
     inner class GetUserByEmail {
         @Test
-        fun `When a user is found, then the correct user is returned`() = runTest {
+        fun `When a user is found, then a successful result with the correct user is returned`() = runTest {
```
Maybe something like "When a user with the given email ..." or so? To distinguish the tests from the `GetUserById` tests or do you think it is unnecessary? (then only resolve this conversation on your own)

Slartibartfass2 replied:
I added "by their ID/email" to distinguish these two very similar tests.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/repository/UserTableRepoTest.kt:
```diff
@@ -289,57 +298,62 @@ class UserTableRepoTest : RepositoryTest(arrayOf(UserTable)) {
-        fun `When a user is not found, then an exception is thrown`() = runTest {
-            assertThrows<NotFoundException> { repo.getUserById(UUID.randomUUID()) }
+        fun `When a user is not found, then no exception is thrown`() = runTest {
+            assertDoesNotThrow { repo.softDeleteUser(UUID.randomUUID()) }
```
Actually this is not the correct behavior regarding to the API documentation?

<img width="1122" height="367" alt="Bildschirmfoto 2025-09-15 um 16 26 59" src="https://github.com/user-attachments/assets/dbc1ff0e-0006-4be1-a0d1-5e1da77fd927" />

Slartibartfass2 replied:
It is. The check is made in `UserService`.

Merseleo replied:
Ah ok, somewhat confusing.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/repository/UserTableRepoTest.kt:
```diff
@@ -289,57 +299,61 @@ class UserTableRepoTest : RepositoryTest(arrayOf(UserTable)) {
     inner class GetUserSettings {
         @Test
-        fun `When a user is found, then the user settings are returned`() = runTest {
+        fun `When a user is found, then the a successful result with the user settings is returned`() = runTest {
```
```suggestion
        fun `When a user is found, then a successful result with the user settings is returned`() = runTest {
```

Slartibartfass2 replied:
Great catch.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt:
```diff
@@ -135,7 +137,7 @@ class CriterionTableRepoTest : RepositoryTest(arrayOf(CriterionTable, ProjectTab
                         .build()
 
-                assertThrows<NotFoundException> { repo.createCriterion(request, testUserId) }
+                assertThrows<SQLException> { repo.createCriterion(request, testUserId) }
```
I think the problem now is, that we throw a SQLException which is not encapsulated as Result type, i.e. not checked in the service layer but only thrown. However, our validation layer does not catch this exception as well and we wouldn't even know the reason for the SQLException, so this seems not the best solution to me. Either we should use the Result type as well or we should still use this `getEntityById` helpers.

Merseleo replied:
Actually this isn't the biggest problem for now, as we actually always check access rights and whether something exists etc. before we retrieve something from the repo layer, but I am not sure whether this is the most elegant way.

Slartibartfass2 replied:
I agree. All of these tests where a `SQLException` is thrown, we actually test whether our DB design prohibits invalid actions. Do you have an idea how we should move forward?

Merseleo replied:
Yeah, actually this isn't the worst idea as we then test whether our constraints are ok. So I think we can keep these tests, however, maybe it could be advantageous to catch such errors in the exception interceptor and log a warning the backend or so in case it happens so we notice that there could be a bug in the backend? Otherwise I would just ignore them for now and think about these tests when we refactor the access rights.

Slartibartfass2 replied:
We don't ignore any exception. Every unexpected exception is handled in the ExceptionInterceptor and mapped to an `INTERNAL` status code while the exception is logged.

Merseleo replied:
Ah ok, perfect.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/criterion/GetCriterionByIdTest.kt:
```diff
@@ -134,14 +134,14 @@ class GetCriterionByIdTest : MainServiceTest() {
 
     @Test
-    fun `When an error occurs while the criterion is retrieved, then an exception is thrown`() = runTest {
+    fun `When an error occurs while the criterion is retrieved, then a TestSpecificException is thrown`() = runTest {
```
Why did you keep this test but removed it e.g. in the UpdateCriterionTest file?

Slartibartfass2 replied:
In `UpdateCriterionByIdTest`, I removed the test where `updateCriterion` throws an exception. Here it's `getCriterionById`. But there's no test for when `getCriterionById` fails. I added it.

Merseleo replied:
Ok, I think I understand.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/paper/GetBackwardReferencedPapersTest.kt:
```diff
@@ -6,124 +6,66 @@ import org.junit.jupiter.api.Test
+        .setId(paperId.toString())
         .build()
 
-    fun failingFunctions(): Stream<Arguments?> = Stream.of(
```
So we drop the mocking of the happy path entirely now or is it too late for me to understand why you deleted it?
And if we delete it, it is possible to use the default lifecycle for the tests again, right? (same applies for all tests in the paper service)

Slartibartfass2 replied:
Because none of these methods return an exception. This way, there's no unhappy path.

Slartibartfass2 replied:
What's wrong with the lifycyle? Did I miss anything?

Merseleo replied:
Ok, so we only keep track of failing functions that can return a Result.failure?!

That we actually started to change the lifecycle to PER_METHOD or so and Dominik changed it back because of his path happy mocking.

Slartibartfass2 replied:
Yes.

I changed the files back.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/paper/GetPaperByIdTest.kt:
```diff
@@ -27,59 +22,31 @@ class GetPaperByIdTest : MainServiceTest() {
         val request = Base.Id.newBuilder().setId("invalid-uuid").build()
 
         assertThrows<InvalidIdException> { mainService.getPaperById(request) }
     }
```
This is only tested here, but not for example in the GetBackwardReferencedPapersTest, for purpose?

Slartibartfass2 replied:
Some included it, some did not. I removed all occurrences of these tests. Since we check ID validity in the validation, I think it's fine to assume their correctness. (We wouldn't even discuss this when we would use a REST API)

Merseleo replied:
Ok thank you. And I agree, that we should assume the ID to be correct. Otherwise we shouldn't for example assume that the field mask is correct as well ...

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/paper/GetPaperByIdTest.kt:
```diff
@@ -27,59 +22,31 @@ class GetPaperByIdTest : MainServiceTest() {
-            mainService.getPaperById(getExampleRequest())
-        }
+    @Test
+    fun `When fetching the paper fails, then a TestSpecificException is thrown`() = runTest {
```
No actually then we get a NotFoundException? And I think we agreed on not using any().

Slartibartfass2 replied:
So far, we always used the `TestSpecificException` for testing the failure of a function. I used a specific matcher instead.

Merseleo replied:
Hm ok.

Merseleo reviewed (changes requested):
Okay, I think I need to take a break for now. Even though it's very easy to review commit by commit thanks to the structure and description, it's still a lot of work and somewhat exhausting. Nevertheless, I can already say: very good work, and considering the size of the refactoring, there are very few points for discussion. 

A few general points: despite the coverage, I think some tests are missing in some places, or the tests no longer match the API documentation, but we shouldn't resolve that here. In general, I see a problem with the SQLExceptions that are not encapsulated, and EntityNotPersisted and InvalidId can also be thrown without this being mentioned anywhere. I'm not quite sure how to deal with this, and perhaps we could discuss it again tomorrow.

Slartibartfass2 committed:
docs: update outdated/bad comments

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ServiceHelper.kt:
```diff
@@ -0,0 +1,13 @@
+import se.uulm.snowballr.backend.repository.IUserTableRepo
+
+/**
+ * Fetches the current user using the [userRepo] and executes the [block] with the current user as parameter.
```
This is an example where (imo) it could be helpful to add a comment that this helper can throw an exception (I think not found or so), in case the current user cannot be fetched? What do you think?

Slartibartfass2 replied:
Yes, I agree. I updated the comment.

Slartibartfass2 committed:
refactor: make getEntityByIdAsResult helper more generic

Now it's getEntityByKeyAsResult and works with every type of key/identifier.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
Actually you can use the `withUser` helper for the `getInviteCandidates` service call as well, right?

Slartibartfass2 replied:
Yes, that's true. I probably forgot to change it there.

Slartibartfass2 committed:
test: add test for when getCriterionById returns an exception

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -324,29 +324,26 @@ class ProjectService(
+    ): GrpcProject.List = withUser(userRepo) { currentUser ->
         val requestedUserId = parseUUID(request.id, EntityType.USER)
-        authorizeAccessTo(requestedUserId, userRepo, AccessType.READ)
+        authorizeAccessTo(currentUser, requestedUserId, userRepo, AccessType.READ)
```
Why do you pass the currentUser to this helper and do not use the `withUser` directly in the `authorizeAccessTo` function?

Slartibartfass2 replied:
No special reason. `authorizeAccessTo` is only used once, but it seems like this could be used more often. Maybe there's some potential. In general, I think the withUser should be directly used in the service method because if two helpers were to use `withUser`, we would request the user twice.

Merseleo replied:
The second argument makes sense 👍, I understand the idea.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/MainServiceTest.kt:
```diff
@@ -173,4 +176,12 @@ open class MainServiceTest : KoinTest {
         clearAllMocks()
         stopKoin()
     }
+
+    /**
+     * Mock the current user that is passed through the [withUser] helper.
+     */
+    protected fun mockCurrentUser(currentUser: User) {
+        every { GrpcContext.getUserIdFromContext() } returns currentUser.id
+        coEvery { userRepoMock.getUserById(currentUser.id) } returns Result.success(currentUser)
+    }
```
Wouldn't it be better to move this into its own file like `ServiceTestHelper` in the `utils` package because I can image that there will follow further mock helpers for the service tests.

Slartibartfass2 replied:
Then we couldn't use the `userRepoMock` and would need to pass it every time. This seems unnecessary to me.

Merseleo replied:
Ah, because our Test classes only extend the MainServerTest, right? Unlucky ...

Slartibartfass2 replied:
Yes, but I think that's fine.

Slartibartfass2 committed:
test: remove tests that test what happens when parseUUID throws

We test ID validity in the validation layer. Testing these cases for all services results in a lot of unnecessary tests.

Slartibartfass2 committed:
test: use specific argument matcher instead of any()

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetAllArchivedProjectsForUserTest.kt:
```diff
@@ -21,63 +19,46 @@ class GetAllArchivedProjectsForUserTest : MainServiceTest() {
 
     @Test
-    fun `When parsing the ID fails, then an exception is thrown`() = runTest {
+    fun `When parsing the ID fails, then an InvalidIdException is thrown`() = runTest {
```
This is another thing that we sometimes test, sometimes not. Maybe we have to discuss this tomorrow again.

Merseleo replied:
But if we keep this test, I think mocking the current user is not necessary as we should fail even before the current user is retrieved (same for related service call tests).

Slartibartfass2 replied:
I removed the parsing tests in 2b21eb5ae0e9d3b99c6ead352246556f82608f58.

Slartibartfass2 reviewed (commented):
Thank you for the work that you put into reviewing this.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/GetCurrentUserTest.kt:
```diff
@@ -1,39 +1,17 @@
-import java.util.UUID
 
 class GetCurrentUserTest : MainServiceTest() {
-    @Test
-    fun `When retrieving the user context fails, then an exception is thrown`() = runTest {
-        every { GrpcContext.getUserIdFromContext() } throws TestSpecificException()
-
-        assertThrows<TestSpecificException> { mainService.getCurrentUser() }
-    }
-
-    @Test
-    fun `When retrieving the user by id fails, then an exception is thrown`() = runTest {
-        val userId = UUID.randomUUID()
-        every { GrpcContext.getUserIdFromContext() } returns userId
-        coEvery { userRepoMock.getUserById(GrpcContext.getUserIdFromContext()) } throws TestSpecificException()
-
-        assertThrows<TestSpecificException> { mainService.getCurrentUser() }
-    }
-
```
Is it clever to delete these tests? Actually they are important to test that the mock works correctly, aren't they?

Slartibartfass2 replied:
I don't see why. The only thing that is used in `getCurrentUser` is `toGrpcUser`. `withUser` is tested separately.

Merseleo replied:
Yeah, i recognized this on my own today.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/SoftDeleteUserTest.kt:
```diff
@@ -22,71 +20,51 @@ class SoftDeleteUserTest : MainServiceTest() {
         coEvery { userRepoMock.getUserById(requestedUserId) } returns Result.success(userToDelete)
 
-        assertThrows<UnauthorizedException.Single> { mainService.softDeleteUser(getExampleRequest()) }
+        assertThrows<UnauthorizedException> { mainService.softDeleteUser(getExampleRequest()) }
```
One thing I noticed multiple times now: Why do you replace the more specific exception by a more general? What if we change the thrown exception from Single to All? Then we have another exception, but the test wouldn't fail?!

Slartibartfass2 replied:
Because there's no difference in the logic. We only have these subtypes to make the exception message easier to configure. For both subtypes, an "unauthorized" status code is removed, which is what this test is about.

Merseleo replied:
Ok, i understand.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/AddPaperToProjectTest.kt:
I think I didn't quite understand the system by which you decided to remove the failAt steps and which should be kept? Can you explain this a bit more? And maybe we should then document this system somewhere for further tests, at least tomorrow in our meeting.

Slartibartfass2 replied:
If there were more than one unhappy paths, I kept the `mockHappyPathUntil` method. Otherwise, it didn't make sense to me. If there are none, this is obsolete. If there's one, it's alright to just write two tests, which are essentially shorter.

Merseleo replied:
Makes sense, I think I understood the system now.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetProjectPaperByRelativeIdTest.kt:
```diff
@@ -146,27 +105,17 @@ class GetProjectPaperByRelativeIdTest : MainServiceTest() {
-                mainService.getProjectPaperByRelativeId(getExampleRequest())
-            }
-        }
-
```
Why did you delete this test? Isn't it important?

Slartibartfass2 replied:
No, `doesProjectExistById` always returns a boolean value. The happy path contains the case where the method returns true and the last test tests when it returns false.

Merseleo replied:
Ah sorry, missed the last test.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/GetUserByEmailTest.kt:
```diff
@@ -23,7 +23,7 @@ class GetUserByEmailTest : MainServiceTest() {
 
         mockCurrentUser(currentUser)
-        coEvery { userRepoMock.getUserByEmail(any()) } throws TestSpecificException()
+        coEvery { userRepoMock.getUserByEmail(any()) } returns Result.failure(TestSpecificException())
```
I think any() can be replaced by exampleEmail here, right?

Slartibartfass2 replied:
Let's do that in https://github.com/SE-UUlm/snowballr-backend/issues/254. I changed occurrences while refactoring, but wouldn't change all cases in this PR.

Merseleo replied:
Ok, sorry I forgot that we have this issue.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/VerifyEmailTest.kt:
```diff
@@ -16,15 +16,15 @@ import java.time.OffsetDateTime
 class VerifyEmailTest : MainServiceTest() {
     @Test
-    fun `When the verification token is not found, then an exception is thrown`() = runTest {
+    fun `When the verification token is not found, then a TestSpecificException is thrown`() = runTest {
```
This is not correct, as a VerificationTokenNotFoundException is thrown (same for other tests in this file).

Slartibartfass2 replied:
Good catch. I updated these test names.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetProjectPaperByRelativeIdTest.kt:
```diff
@@ -138,4 +139,13 @@ class GetProjectPaperByRelativeIdTest : MainServiceTest() {
     }
+
+    @Test
+    fun `When a non existing project is requested, then a NotFoundException is thrown`() = runTest {
```
Sometimes it is "non existing" and sometimes "non-existing", maybe we should stick to one version?

Slartibartfass2 replied:
Actually, it's nonexistent. I changed all occurrences.

Merseleo replied:
Ok, I think existing would be actually fine too, but probably both versions are correct.

Merseleo reviewed (changes requested):
Done! I've looked at all the commits and tested some of the calls again, but it's not possible to do everything (in terms of time). I hope I haven't overlooked anything. After you've made your changes, or at least after my second review, I'll go through everything file by file to see if anything is missing or anything like that, but I think the refactoring is fine overall.  Thanks for your work.

Slartibartfass2 committed:
refactor: use withUser for getInviteCandidates

Slartibartfass2 committed:
refactor: add ensureCurrentUserIsProjectMember

Slartibartfass2 committed:
docs: change (non-)existing to (non)existent

The latter is the correct english form.

Slartibartfass2 reviewed (commented):
Thank you for your time.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -265,26 +263,10 @@ class UserTableRepo(
-        } else {
-            Result.failure(NotFoundException(EntityType.USER, email, identifierType = IdentifierType.EMAIL))
-        }
+        getEntityByKeyAsResult(::getPasswordHashByUserMailOrNull, EntityType.USER, email)
```
Is it possible you forgot the `IdentifierType.EMAIL` here?

Slartibartfass2 replied:
Good catch!

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/RepoResultHelper.kt:
```diff
@@ -1,44 +1,57 @@
  * returned, containing a [NotFoundException].
  *
- * @param T The type of the entity.
+ * @param Key The type of the key.
```
Maybe `KeyType` or so is better?

Slartibartfass2 replied:
I renamed it `KeyT` similar to `EntT`.

Merseleo replied:
Great, thank you!

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -150,27 +149,26 @@ class UserTableRepo(
+        mapper = { row -> row[UserTable.passwordHash] },
+        select = UserTable.select(UserTable.passwordHash),
+        where = { UserTable.email eq email },
+    )
```
I am not sure whether this is really better than just creating the normal SQL statement. Because now your helper is exactly the normal SELECT FROM WHERE statement. Why no create a lambda or so with this statement for the case and leave the `getEntityOrNull` helper?

Slartibartfass2 replied:
I agree; I changed it back.

Merseleo replied:
But you forgot to change back the `getEntities` function? Maybe you can adapt the Commit 14002780e329c115430c685e0b8b96ad5aff4d72? (or @mowi12)

mowi12 replied:
Done

Merseleo reviewed (changes requested):
Ok, I think there are only minor things left. You can enable auto-merge! After your responses I will have a look at all files again and check the repository functions again, but I dont think there will be any things to discuss and then I can automatically merge when I am finished. In case I really find something, maybe @mowi12 can work on the necessary changes or so.

Slartibartfass2 committed:
chore: remove unused test annotations

Slartibartfass2 committed:
refactor: rename Key to KeyT

Slartibartfass2 committed:
chore: fix minor issues

Slartibartfass2 reviewed (commented):
Thank you for the review.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/VerificationTokenTableRepo.kt:
```diff
@@ -49,11 +48,7 @@ class VerificationTokenTableRepo(
-            .where { VerificationTokenTable.token eq token }
-            .map { it.toVerificationToken() }
-            .singleOrNull()
+        VerificationTokenTable.getEntityOrNull(ResultRow::toVerificationToken) { VerificationTokenTable.token eq token }
```
I think this method could return a Result as well instead of returning null in the error case which leads to throwing an exception on the service layer. What do you think?

mowi12 replied:
Yes, and no.

If I understood it correctly, the result thing would only return a `NotFoundException`.
If used the new method with `getOrThrow` and still want to return a `VerificationTokenNotFound` exception, e would need to wrap this in a `try`-`catch` block, which would only bloat the code.

Since this function is only used here, I think this is okay.

Merseleo replied:
Actually no, then you misunderstood the ResultType. You can wrap every exception you want.

```kotlin
override suspend fun getVerificationTokenByValue(token: String): Result<VerificationToken> = db.query {
        val token = VerificationTokenTable.getEntityOrNull(ResultRow::toVerificationToken) { VerificationTokenTable.token eq token }

        return@query if (token != null) {
            Result.success(token)
        } else {
            Result.failure(SnowballRException.VerificationTokenNotFoundException())
        }
        
    }
```

This is an example how this function then could look like. Do you think this refactoring would improve the code or not? Or should be keep it for now and discuss such a scenario in another issue, as we will probably only use this function once?!

mowi12 replied:
No, I didn't misunderstood the Result Type. But I was so focused on refactoring the method using the new helper methods (which throw NotFoundException), that I didn't think about simply doing it directly.

Either way this doesn't really improve the code just moves the focus into a different section.
You can write this down and we will discuss this later, but for now I think we can keep it like it is.

Merseleo replied:
All right, that's how it's done.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepo.kt:
```diff
@@ -120,31 +122,30 @@ class ProjectPaperTableRepo(
+                .selectAll()
+                .where(where)
+                .map { it.toProjectPaper() }
+                .singleOrNull()
```
I think the `getEntityOrNull` helper can be used here?

mowi12 replied:
I think since the `where` is a more complex expression (it contains an `and`) the type is not matching. At least I didn't get it to work.

Merseleo replied:
```kotlin
val projectPaper2 = ProjectPaperTable.getEntityOrNull(ResultRow::toProjectPaper) {
                (ProjectPaperTable.projectId eq projectId) and (ProjectPaperTable.localPaperId eq relativeId)
            }
```

This would be an option.

mowi12 replied:
You're right. This looks good!

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/paper/GetPaperByIdTest.kt:
```diff
@@ -3,83 +3,40 @@ package se.uulm.snowballr.backend.service.paper
-        }
-        coEvery { paperRepoMock.getPaperById(paper.id) } returns paper
+    @Test
+    fun `When a paper is retrieved, then no exception is thrown`() = runTest {
```
```suggestion
    fun `When a paper is retrieved successfully, then no exception is thrown`() = runTest {
```

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetAllArchivedProjectsForUserTest.kt:
```diff
@@ -1,98 +1,56 @@
 
     @Test
-    fun `When retrieving requested user fails, then an exception is thrown`() = runTest {
+    fun `When retrieving requested user fails, then a TestSpecificException is thrown`() = runTest {
```
```suggestion
    fun `When retrieving the requested user fails, then a TestSpecificException is thrown`() = runTest {
```

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetAllDeletedProjectsForUserTest.kt:
```diff
@@ -1,94 +1,56 @@
 
     @Test
-    fun `When retrieving requested user fails, then an exception is thrown`() = runTest {
+    fun `When retrieving requested user fails, then a TestSpecificException is thrown`() = runTest {
```
```suggestion
    fun `When retrieving the requested user fails, then a TestSpecificException is thrown`() = runTest {
```

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/RegisterTest.kt:
```diff
@@ -9,58 +9,21 @@ import org.junit.jupiter.api.Test
-        coEvery { emailManagerMock.sendVerificationEmail(any(), any()) } throws TestSpecificException()
-
-        assertThrows<TestSpecificException> { mainService.register(request) }
-    }
```
Isn't this test important? Because actually the sendEmail function can throw an error and maybe we should wrap this into a result as well?

mowi12 replied:
Imo you're right. This test has nothing to do with the `saveVerificationToken()` function, since the sending of the mail can fail. I will add this back.

Merseleo replied:
Ok thank you. Maybe we should later add a comment to the `sendVerificationEmail` that it can fail, but this will be done later as we have to adapt the function anyway when we remove unnecessary `any()` occurrences. But for now, I think this is ok this way.

Merseleo reviewed (changes requested):
Ok, I found two functions where I am not certain, whether it wouldn't be better to wrap the result in our ResultType, maybe @mowi12 have more expertise regarding these functions. The remaining comments are minor things, further problems I found will be resolved in another PR.

mowi12 committed:
chore: update wording

mowi12 reviewed (commented):
I think I updated everything.

Merseleo reviewed (changes requested):
Ok i made two suggestions that show what were my thoughts in the review. I would like to know your opinion on that, but if you are not sure about how we should deal with that, we can ignore both conversations, I add them to the list of things that should be fixed in general in the backend repository and then we discuss it later.

mowi12 committed:
refactor: use new helper function to simplify code

Merseleo reviewed (approved):
Ok thank you @mowi12 and especially @Slartibartfass2 for all the effort you put into this PR. I think this helps a lot and we improved the code quality a lot. Probably there are still some problems or open discussions, but it can be fixed later or it isn't important.
